plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextEditorCommands.java

```diff
@@ -127,12 +134,75 @@ protected void doExecute() {
             }
 
             instanceRole.setRepresentedInstance((AbstractInstance) capellaParticipant);
-            instanceRoles.add(instanceRole);
+            instanceRoles.add(instanceRole);            
           }
         }
+        
+        removeParticipantsFromDiagram(instanceRoles, participants);
+             
+        if (switchParticipants(instanceRoles, participants)) {
+        	syncGraphicalOrdering();
+        } 
       }
     });
   }
+  
+  /**
+  If a participant is in diagram, but not in editor, delete it
+  @param instance roles from diagram
+  @return participants from editor
+  @throws what kind of exception does this method throw
+  */
+  private static void removeParticipantsFromDiagram(List<InstanceRole> instanceRoles, EList<Participant> participants) {
+	  List<InstanceRole> irToRemove = new ArrayList<InstanceRole>();
+      for (InstanceRole ir : instanceRoles) {
+      	List<String> participantsName = participants.stream().map(x -> x.getName()).collect(Collectors.toList());
+      	if (!participantsName.contains(ir.getName())){       		
+      		irToRemove.add(ir);       		
+      	}
+      }
+      for (InstanceRole ir : irToRemove) {
+      	instanceRoles.remove(ir);      	
+      }
+  }
+  
+  
+  /**
+  Switch participants in diagram if they are not in the same order as they are in editor
+  @param instance roles from diagram
+  @return true if the order was changed, false if not
+  */
+  private static boolean switchParticipants(List<InstanceRole> instanceRoles, EList<Participant> participants) {
+	  boolean updateNeeded = false;
+      for (int i = 0; i < instanceRoles.size(); i++) {       	
+      	if (!instanceRoles.get(i).getName().equals(participants.get(i).getName())) {
+      		for (int j = 0; j < participants.size(); j++) {
+      			if (instanceRoles.get(i).getName().equals(participants.get(j).getName())) {
+      				((EList<InstanceRole>) instanceRoles).move(j, instanceRoles.get(i));  
+      				updateNeeded = true;     				
+      			}
+      		}
+      	}
+      }
+      return updateNeeded;
+  }
+  
+  /**
+  Synchronize graphical ordering in diagram
+  */
+  private static void syncGraphicalOrdering() {
+	  DDiagram dDiagram = EmbeddedEditorInstance.getDDiagram();       
+  	EList<AnnotationEntry> ownedAnnotationEntries = dDiagram.getOwnedAnnotationEntries();
+      EObject data = null;
+      for (AnnotationEntry annotationEntry : ownedAnnotationEntries) {
+        if ((annotationEntry != null) && (annotationEntry.getData() instanceof Diagram)) {
+          data = annotationEntry.getData();
+        }
+      }     
+      Diagram diagram = (Diagram) data;       
+      AbstractModelChangeOperation<Boolean> synchronizeGraphicalOrderingOperation = new SynchronizeGraphicalOrderingOperation(diagram, true);
```

Comment:
can remove var diagram and directly cast in new SynchronizeGraphicalOrderingOperation